pyhap/characteristic.py

```diff
@@ -125,12 +130,18 @@ def to_valid_value(self, value):
         """Perform validation and conversion to valid value"""
         if self.properties.get('ValidValues'):
             if value not in self.properties['ValidValues'].values():
-                raise ValueError('%s is not a valid value'.format(value))
+                logger.error('%s: value=%s is an invalid value.',
```

Comment:
Sorry, didn't my comment in quick enough before.

It seems really strange to handle the error here. Surely it's better just to raise an Exception and have the caller handle it. Doing this means we now have special return values to indicate a failure.

If people don't like the traceback, then they should be catching the `ValueError` and logging it explicitly. (We could modify the examples to do this).

Addtionally if the `ValueError` is caused by iOS, then we should be returning the correct status code to iOS to indicate an invalid value has been sent.

Reply:
I wouldn't catch in `set_value` but in whatever called `set_value`. It seems strange to rob the caller of the ability to handle the error, just because our one usage now dosen't require that ability, and tracebacks are ugly.

Reply:
Also returning `None` hides the error from the caller. Making even less obvious to the developer using the API that something has gone wrong.

Submitting an invalid value the perfect time to fail fast and fail hard. Ensuring that anything that might rely on that valuing be set won't run unless the failure is explicitly handled. 

Doing anything else if just going to quietly hide the error in impenetrable logs.

Reply:
I'm thinking from a general standpoint here. You should configure your accessory the way that only valid values are set. If for some reason a value isn't valid, you should get a notification with the invalid value so you can update your program.
It would be a waist of resources to write the error handling for every set_value call individually. Think of the Home Assistant `Thermostat` implementation. It alone uses 11 set value calls.

Reply:
By your own argument then, there is no reason not to raise an exception. If they only happen rarely, and always indicate that a serious problem has occurred, then they should just be allowed to peculate to the top. Doesn't matter how the traceback looks, it should never happen.

You don't need to do error handling on every call to `set_value`, only on calls where you are pulling values from somewhere that you don't control, and thus can't guarantee will always be in bounds.

Additionally we can add some smarter error handling at the top of the accessory to capture these exceptions, and restart the appropriate accessory loop if people want that.

Reply:
Go ahead, but I'm frustrated if I see those error logs and should try to find the issue: https://github.com/home-assistant/home-assistant/issues/13676

At least for Home Assistant you can't never be sure that the correct values are supplied and there will always be one you haven't caught.

Reply:
I understand that, but the solution isn't to remove functionality from py-HAP. We should modify the accessory components when we integrate with HA to capture the exceptions. This could simply be done by creating a `set_value_log_errors()` helper in HA, or on the accessory.

This way we also get the added benefit of being able to write HA specific messages, which can pull in `entity_id` and component name. Giving us the debug info needed to fix components.

Reply:
Of course that is possible. I just think it defies the purpose. A user shouldn't ever get tracebacks. If an error occurs it should be handled as error log message.

Reply:
It would be different if `HAP-python` would just be an API the user doesn't interact with directly, but it isn't.

Reply:
I agree that a user shouldn’t get trace backs. But it’s on HA to catch Exceptions, it’s not for py-HAP to accept any old input and carry on, just so HA doesn’t have do proper error handling. 

py-HAP is just an API, a HA user shouldn’t have to interact directly with it. The HA ui should be doing all the heavy lifting. 

If people are using py-HAP directly then it’s on them to catch errors and handle them, and they should be devs and know what to do with them. 

If people are just using the demo accessories then they should be updated to capture the exception and produce a meaningful error message. 

In each these scenarios the output will be different, we should be picking one and forcing everyone to use it. 

And specifically with HA, doing the error handling in HA, rather than in py-hap means we can produce much better error messages(including HA info). Helping us debug issues quicker.

Reply:
Can I also point out that your proposed error message is very unhelpful. It says there is an invalid value, be not what caused it. Which accessory tried to set the value, and onto what characteristic, and why is it invalid?

Reply:
One final point. If your frustrated by the amount of work needed to do the error handling. I’m happy to do that work, both here and in HA.

Reply:
I disagree. `'%s: value=%s is an invalid value.', self.display_name, value` should cover almost all cases. It's true that you don't get the accessory, but you know it already since it hasen't been updated. You also get the distinction between `valid values` and `numerics`. There isn't really anything more I would add / need.

---

> One final point. If your frustrated by the amount of work needed to do the error handling. I’m happy to do that work, both here and in HA.

I won't stop you from looking through the open issues and suggesting solutions.

Reply:
I think we are overthinking this a bit. It should be very rare to get invalid values:
- iOS cannot send an invalid value, because it picks values from the min/max/validValues fields that the characteristic defines. In the worst case, even if Apple change something, your accessory will not get added to the Home app in the first place (I tried)
- Accessories in HAP are developed from devs not end users. This is the target "audience" so to speak. Thus, we need the best solution for devs to be able to pinpoint the problem.

That being said, I think that the current ValueError is good enough. When the dev is coding a new Accessory, he/she will probably get these early in testing. Once they are gone, they should be gone for good.